README.md

```diff
@@ -1,77 +1,91 @@
 # ZKIR: Zero-Knowledge Intermediate Representation
 
-ZKIR is a set of high-level operations (HLOs) designed for zero-knowledge (ZK)
-provers, inspired by [StableHLO](https://github.com/openxla/stablehlo). It
-leverages
-[MLIR (Multi-Level Intermediate Representation)](https://mlir.llvm.org/) to
-enable domain-specific optimizations and support for heterogeneous proving
-backends.
+**ZKIR** is an Intermediate Representation (IR) designed specifically for
+Zero-Knowledge (ZK) proving, built on top of
+[MLIR (Multi-Level Intermediate Representation)](https://mlir.llvm.org/). The
+core goal of ZKIR is to enable domain-specific optimizations and efficiently
+support diverse, heterogeneous proving backends.
 
 ## Motivation
 
-Zero-knowledge (ZK) proof systems often require domain-specific arithmetic
-operations—such as field multiplication and modular inversion that are difficult
-to express and optimize at low-level IRs like LLVM IR.
+ZK proving systems often require domain-specific arithmetic operations, such as
+field multiplication or modular inversion. These operations are difficult to
+express and optimize at the level of low-level IRs like LLVM IR.
 
 ### Why not use LLVM IR?
 
 LLVM IR is designed primarily for general-purpose CPU targets, which introduces
 several limitations:
 
-- **Lack of domain semantics**: Field operations like `-x` are reduced to
-  generic integer operations. If the modulus is 11, the code will be as follows:
+- **Lack of domain semantics**: Field operations like $-x$ are reduced to
+  generic integer operations. If $x$ is non-zero and the modulus is 11, the code
+  becomes:
 
   ```mlir
    %0 = llvm.mlir.constant(11 : i32) : i32
    %1 = llvm.sub %0, %arg0 : i32
   ```
 
-  This makes algebraic simplification and reasoning nearly impossible.
+  We know that negation is an involution operation ($-(-x) = x$), but we lose
+  this meaning when the operation is expressed in LLVM IR. This makes algebraic
+  simplification impossible.
 
-- **Loss of intent**: Without explicit semantic representation, higher-level
-  optimizations or rewrites (e.g., canonicalizing double negations) become
-  infeasible.
+- **Loss of hardware abstraction**: LLVM IR cannot be compiled for GPUs or TPUs.
```

Comment:
LLVM IR can be compiled to GPU I think. https://releases.llvm.org/17.0.1/docs/NVPTXUsage.html

> To support GPU programming, the NVPTX back-end supports a subset of LLVM IR along with a defined set of conventions used to represent GPU programming concepts. This document provides an overview of the general usage of the back- end, including a description of the conventions used and the set of accepted LLVM IR.